### DISCUSSION TITLE
modify the function EnableToolPorts() in vtkPlusNDITracker

### DISCUSSION THREAD

**Discussion by acc2024** [PlusToolkit/PlusLib#1148] [Q&A] [answered]
Hi!
I am reviewing the code the for the device `vtkPlusNDITracker`. At the end of the` InternalUpdate()` function, there is a call to the function `EnableToolPorts()` to "Make the newly connected tools available". Can anyone summarize how we may adapt this function for a non-NDI tracker? I am not sure to understand which Plus function that I need to call to make a tool available. I would appreciate your help/comments on the issue. 
Thank you!

**Comment by lassoan** [comment, marked as answer]
This is an extra required for NDI trackers to configure tools. Other trackers typically don't need it, because tools are configured when the connection to the tracker is established.

**Reply by acc2024** [reply]
Hi Andras, thank you for your prompt answer and clarification! I have a follow up question: how to update the status (MISSING or OK) in fCal from the device file?

I added the following code at the end of the of the `InternalUpdate()` function:

```
ToolStatus toolFlags = TOOL_OK; //TOOL_MISSING;
this->ToolTimeStampedUpdate("ReferenceToTracker", toolTransform, toolFlags, this->FrameNumber, toolTimestamp);
```

But it doesn't seem to work. I added more details about this issue in the discussion: #1147 

Thank you!

**Reply by adamrankin** [reply]
Tool flags is the correct way to set the status of a transform in the tool's buffer. When you say device file, do you mean saved data, or in the fCal tool state display?

**Reply by acc2024** [reply]
Hi Adam, Thank you for your reply! By device file I mean a new plus device (vtkPlusMyTracker.cxx/h) similar to vtkPlusNDITracker. The transform is computed in vtkPlusMyTracker.cxx using pos and orientations that I get from an EM tracker. Then I use the code below  in the `InternalUpdate()`  function:
```
ToolStatus toolFlags = TOOL_OK; //TOOL_MISSING;
this->ToolTimeStampedUpdate("ReferenceToTracker", toolTransform, toolFlags, this->FrameNumber, toolTimestamp);
```

The aim is to perform a stylus calibration using fCal. Currently the button "Start" under the tab "Stlus Calibration" is disabled even though I can see that the  pos and orientations values from the three sensors are correctly displayed in the fCacl log file. 

Here is a snippet of my configuration file: 

```
<Device
      Id="TrackerDevice" 
      Type="MyTracker"
      AcquisitionRate="5"
      ToolReferenceFrame="Tracker"
	  IPAddress= "xx.xx.xx"> 
      CheckDSR="true">
      <DataSources>
        <DataSource Type="Tool" Id="Reference" BufferSize="1500" PortName="0" AveragedItemsForFiltering="20">
          <CustomProperties>
 ...
 </DataSource>
        <DataSource Type="Tool" Id="Probe" BufferSize="1500" PortName="1" AveragedItemsForFiltering="20">
          <CustomProperties>
            <Manufacturer>
</DataSource>
        <DataSource Type="Tool" Id="Stylus" BufferSize="1500" PortName="2" AveragedItemsForFiltering="20">


...
 </DataSource>
      </DataSources>
      <OutputChannels>
        <OutputChannel Id="TrackerStream">
          <DataSource Id="Reference" />
          <DataSource Id="Probe" />
          <DataSource Id="Stylus" />
        </OutputChannel>
      </OutputChannels>
      <Parameters />
    </Device>
  </DataCollection>
```

Do I need to do any additional code changes to perform a stylus calibration other than the changes mentioned above to add a new device and a configuration file? 

 I would appreciate your help/comments.

Thank you!

**Reply by adamrankin** [reply]
What does the textbox under Start say?

This code https://github.com/PlusToolkit/PlusApp/blob/master/fCal/Toolboxes/QStylusCalibrationToolbox.cxx shows you the various reasons for the button to be greyed out. It could be something other than the transform state.

**Reply by acc2024** [reply]
Thank you! I will check this file!

Under the tab "Stylus Calibration" it says "StylusTipToStylus transform is absent, calibration needs to be performed."  See the image below for a screenshot: 
<img width="191" alt="Capture_stylus_calibration" src="https://github.com/PlusToolkit/PlusLib/assets/155571573/38320da3-b06b-4de4-8ea9-44e794fc0788">

Thank you!

**Reply by acc2024** [reply]
Hi Adam,
After checking the file, you mentioned above (`QStylusCalibrationToolbox.cxx`), and the log, I found the error/warnings message: `"No transform found between stylus and reference!"` and ` SelectedChannel object is invalid or the selected channel does not contain tracking data!`

Because of the following function calls:
```
IsExistingTransform(
      objectMarkerCoordinateFrame.c_str(), referenceCoordinateFrame.c_str())
```
the values for the arguments are: 
```
objectMarkerCoordinateFrame.c_str() = "Stylus"
referenceCoordinateFrame.c_str() =  "Reference"
```
and
```
if (this->SelectedChannel == NULL || this->SelectedChannel->GetTrackingDataAvailable() == false)
{
  LOG_WARNING("SelectedChannel object is invalid or the selected channel does not contain tracking data!");
  return PLUS_FAIL;
}
```

As a consequence, a flag is set to disable the Start button...  

I am still not sure what is the issue. Please let me know if you have further comments/hints on this...
Thank you!

**Reply by adamrankin** [reply]
Can you confirm that your code is adding StylusToTracker and ReferenceToTracker into their respective buffers?

**Reply by acc2024** [reply]
Hi Adam,

I think yes.  I am using the code below for `InternalUpdate` StylusToTracker, ProbeToTracker, and StylusToTracker.

```
this->ToolTimeStampedUpdate("ReferenceToTracker", toolToTrackerTransform, toolFlags, this->FrameNumber, toolTimestamp);

this->ToolTimeStampedUpdate("ProbeToTracker", toolToTrackerTransform, toolFlags, this->FrameNumber, toolTimestamp);

this->ToolTimeStampedUpdate("StylusToTracker", toolToTrackerTransform, toolFlags, this->FrameNumber, toolTimestamp);
```
To log the pos and orientations I am using LOG_INFO. The values in the logs are correct.

Please let me know if I am missing something. 

Thank you!

**Reply by lassoan** [reply]
Probably something is wrong with your configuration file. For example, if your device provides tools but they are not included in the output stream then they will not appear in the application. Can you copy here the contents of your device set configuration file?

You can also record the transforms to a file to verify that all the transforms and their status are correct.

**Reply by acc2024** [reply]
Hi Andras,
Thank you! I will log the transformations too. Here's the config file for the device (I removed details about the sensor/manufacture that need to be updated). 

```
<Device
      Id="TrackerDevice" 
      Type="MyTracker"
      AcquisitionRate="5"
      ToolReferenceFrame="Tracker"
	  IPAddress= "xx.xx.xx"> 
      CheckDSR="true">
      <DataSources>
        <DataSource Type="Tool" Id="Reference" BufferSize="1500" PortName="0" AveragedItemsForFiltering="20">
          <CustomProperties>
 ...
 </DataSource>
        <DataSource Type="Tool" Id="Probe" BufferSize="1500" PortName="1" AveragedItemsForFiltering="20">
          <CustomProperties>
            <Manufacturer>
</DataSource>
        <DataSource Type="Tool" Id="Stylus" BufferSize="1500" PortName="2" AveragedItemsForFiltering="20">

...
 </DataSource>
      </DataSources>
      <OutputChannels>
        <OutputChannel Id="TrackerStream">
          <DataSource Id="Reference" />
          <DataSource Id="Probe" />
          <DataSource Id="Stylus" />
        </OutputChannel>
      </OutputChannels>
      <Parameters />
 </Device>
```


Thank you!

**Reply by acc2024** [reply]
I checked the values of the transformations that are added using `ToolTimeStampedUpdate`. They look valid, for example in one update: 
transform = {0.00150547 0.999719 -0.0236476 0 2.54325e+79 1.71329e+256 1.39805e-76 8.62243e-319
-0.0959552 0.164198 0.981749 0 2.0614e+272 7.15781e-290 4.07446e-317 5.93137e-309}

Are there any issues with the config file above? It is similar to the config file that I use for ndi which is working correctly in fCal for ndi.
Thank you!

**Reply by adamrankin** [reply]
If you can confirm that your code is definitely putting transforms into the different tool buffers (yes, via `ToolTimeStampedUpdate`, correctness of numbers doesn't matter for now), and that fCal is correctly set to look into a channel that combines those transforms (looks like `TrackerStream` from your config file, I think the drop down is in the top left?), then it should be able to calculate StylusToReference, and let you proceed.

Can you confirm those aspects?

**Reply by acc2024** [reply]
Hi Adam,
Thank you for the review of the issue. Below are the answers/clarification: 
> If you can confirm that your code is definitely putting transforms into the different tool buffers (yes, via `ToolTimeStampedUpdate`, >correctness of numbers doesn't matter for now), 

Acc: In the code before adding the transforms using `ToolTimeStampedUpdate`, I logged their values using the LOG_INFO. I checked the logs, the values are printed for each transform in the log file.  Is this the confirmation that it is needed? 

>and that fCal is correctly set to look into a channel that combines those transforms (looks like `TrackerStream` from your config file, 

Acc: I used the name `TrackerStream` for the output channel in the config file because it was named that way for the ndi config file. This is the only location where I used `TrackerStream`

>I >think the drop down is in the top left?), then it should be able to calculate StylusToReference, and let you proceed.
> 
Acc: I am not sure to understand this question, please see below an image of the tabs for fCac
![image](https://github.com/PlusToolkit/PlusLib/assets/155571573/2a10c885-36b9-45af-9af4-3e9da72551a6)

Please let me know if you need any additional clarification. 
Thank you!

**Reply by acc2024** [reply]
When I run the `TrackingTest.exe` application (source code located in `PlusLib\src\PlusDataCollection\Testing\TrackingTest.cxx` ) with my config file, I get the error 
`Tracking is not enabled!`

However, from the device file (`vtkPlusYourDeviceNameSource.cxx`), in the `ReadConfiguration` function I was able to read all the tracker tools available in my config file using the following code: 

```
ReadConfiguration(vtkXMLDataElement* config) 
{
    XML_FIND_DEVICE_ELEMENT_REQUIRED_FOR_READING(deviceConfig, config);
    XML_READ_CSTRING_ATTRIBUTE_OPTIONAL(IPAddress, deviceConfig);

    XML_FIND_NESTED_ELEMENT_REQUIRED(dataSourcesElement, deviceConfig, "DataSources");
    for (int nestedElementIndex = 0; nestedElementIndex < dataSourcesElement->GetNumberOfNestedElements(); nestedElementIndex++)
    {
	    vtkXMLDataElement* toolDataElement = dataSourcesElement->GetNestedElement(nestedElementIndex);
	    if (STRCASECMP(toolDataElement->GetName(), "DataSource") != 0)
	    {
		    // if this is not a data source element, skip it
		    continue;
	    }
	    bool isEqual(false);
	    if (igsioCommon::XML::SafeCheckAttributeValueInsensitive(*toolDataElement, "Type", vtkPlusDataSource::DATA_SOURCE_TYPE_TOOL_TAG, isEqual) != PLUS_SUCCESS || !isEqual)
	    {
		    // if this is not a Tool element, skip it
		    continue;
	    }
	    const char* toolId = toolDataElement->GetAttribute("Id");
	    if (toolId == NULL)
	    {
		    LOG_ERROR("Failed to initialize Mytool: DataSource Id is missing");
		    continue;
	    }
	    igsioTransformName toolTransformName(toolId, this->GetToolReferenceFrameName());
	    std::string toolSourceId = toolTransformName.GetTransformName();
	    vtkPlusDataSource* trackerTool = NULL;
	    if (this->GetTool(toolSourceId, trackerTool) != PLUS_SUCCESS || trackerTool == NULL)
	    {
		    LOG_ERROR("Failed to get the tool: " << toolSourceId);
		    continue;
	    }
	    else 
	    {
		    LOG_INFO("Mytool: " << toolSourceId);
	    }

...
}

```
The output of this is: 

```
Mytool: ReferenceToTracker
Mytool: ProbeToTracker
Mytool: StylusToTracker
```


Any hint/comments on why I can read the tools from the device file but not from the `TrackingTest.exe`? The test does connect successfully to the device, but it doesn't find the tools. 
Thank you!

**Reply by adamrankin** [reply]
The device has tools, but the channel might not. Does your device have only one channel?

Edit: The `Tracking is not enabled!` error indicates the channel has no tools, but `ReadConfiguration` is checking the device tool list.

**Reply by acc2024** [reply]
Thank you for the clarification! 

Yes, the tracker sensors have multiple channels. Also, I have `TrackerStream` and `VideoStream`. 

In the config file above, I added a device `TrackerVideoDevice`:

```
    <Device Id="TrackedVideoDevice" Type="VirtualMixer">
      <InputChannels>
        <InputChannel Id="TrackerStream" />
        <InputChannel Id="VideoStream" />
      </InputChannels>
      <OutputChannels>
        <OutputChannel Id="TrackedVideoStream" />
      </OutputChannels>
      <Parameters />
    </Device>
```
This does not seem to fix the error `Tracking is not enabled`. Are there other changes that I should make to the config and/or the device files (`vtkPlusYourDeviceNameSource.cxx`) to handle multiple channels?
Thank you!

**Reply by adamrankin** [reply]
TrackingTest just grabs the first channel, are you sure it's the right one? I would test with Slicer to see if your device is properly sending transforms. If it's properly sending transforms, then we can diagnose your device a bit further.
